### ISSUE TITLE
Change raw pointers to unique_ptr for static variables in new ELI

### ISSUE THREAD

**Issue by jjwilke** [sstsimulator/sst-core#394]
This is a general cleanup of memory leaked at the end of the simulation so that valigrind/leak sanitizer are happy

**Comment by jjwilke** [comment]
We need to be careful with this for El Capitan/XC8. There are compiler issues with unique_ptr and static initialization.

**Comment by jjwilke** [comment]
Related to #390. This requires significant modifications to library unloading to avoid segfaults (related to #443) by cleaning things out before we call dlclose.

**Comment by leekillough** [comment]
Also https://github.com/sstsimulator/sst-core/pull/1222

**Comment by leekillough** [comment]
With #1222 , maybe this should be closed, or the description changed?   I have fixed many Valgrind leaks. Unfreed pointers at program exit are not a serious leak IMO, and I don't count them in my Valgrind tests. More serious are when unfreed pointers are modified or go out of scope and no other pointers point to the allocated memory.

The code in ELI which is referred to here seemed to deal with [static initialization order fiasco](https://isocpp.org/wiki/faq/ctors#static-init-order) by creating `static` pointers or `static std::unique_ptr` at namespace/class scope which are allocated and call the constructor, which depends on initialization order, the first time that they are used. But this is overly complicated. A Meyers Singleton, or a static variable local to a function, gets initialized the first time its block is executed. A reference to the static variable is returned by the function and used everywhere the variable is needed elsewhere. [Benchmark tests](https://www.modernescpp.com/index.php/thread-safe-initialization-of-a-singleton/) have shown this is the fastest way to initialize a static-lifetime variable only once in a thread-safe way. #1222 changes the static-pointer-allocated-on-first-use to Meyers singletons by putting the `static` variable (not a pointer) in a function and constructing it on the first call.

When dynamically allocated memory is needed, I prefer `std::unique_ptr` and `std::make_unique` so that the memory is automatically freed and doesn't have to be explicitly `delete`d in a destructor, and is exception-safe (see https://github.com/sstsimulator/sst-core/pull/1255). I would look for all instances of `delete` in destructors, and see whether the freed pointer can be replaced by a `std::unique_ptr`.

**Comment by feldergast** [comment]
I don’t think Meyer’s singletons will work.  The whole point of those initializations is that they MUST run at library load time.  There are no functions to call to have the Meyer’s singleton initialize until after those variables have been initialized. The variable that gets initialized is actually not even really needed, what’s needed is the side effects from the functions that are called at load time that initialize that variable. I suspect you could make this all work with unique_ptr, but Meyer’s singletons will not, unless you can put them in a function that runs at library load time (i.e. when dlopened).

**Comment by leekillough** [comment]
There are function attributes you can use to make functions run automatically at load time.

The raw pointers (or `std::unique_ptr`) in the ELI code were global pointers which were initialized to `nullptr`, possibly only after load-time. The only time that they called the constructors which had the "side effects", was when a function was called and noticed that they were `nullptr` and initialized them to something else by calling `new` and the constructor with side effects. I am saying that this is unnecessary with static variables defined _inside_ functions, even if they are part of a library which gets loaded later.

> The variable that gets initialized is actually not even really needed, what’s needed is the side effects from the functions that are called at load time that initialize that variable.

Which is precisely what happens when a static variable _defined inside a function_ is initialized the first time that that function is called, presumably after library loading.

A Meyers singleton is just a name for a static variable defined _inside_ of a function (not in a class or namespace) so that it is constructed (initialized) the first time the function (or the block inside it containing the static variable) is entered and doesn't suffer from the "initialization order fiasco" of depending on other global variables which might not have been initialized yet. 

I take it from your comments that you want variables to be initialized at load time. What I am saying is that the changes I made to the "ELI" code in that PR, has the same effects as initializing a pointer with a `new type(constructor_args)` call the first time the pointer is needed. The runtime behavior is the same, whether it's a static variable in a function which gets initialized the first time the function is called, or a global variable in a library which gets initialized to `nullptr` on library load, but then gets initialized with `new` and a constructor call the first time a function checks whether it's nullptr and calls `new` and the constructor if it's not.

A static variable _inside a function_ does the same thing. You may not need it after that, but it will be initialized the first time the code executes past the variable definition, and its initialization will have all of the side effects you want. It also will not occur until that function is entered, so that other initializations it depends on will have completed.

Method 1:

```c++
FancyObject* fancy_ptr = nullptr;

void myfunc(ARG args){
    if ( ! fancy_ptr ) fancy_ptr = new FancyObject(args);
   // FancyObject initialized, but we have a pointer laying around
}
```
Method 2:
```c++
void myfunc(ARG args){
   static FancyObject my_object(args);
   // FancyObject initialized; static variable will hold its state but we don't care
}
```
The two behave the same, whether or not they're in a loaded library. The second is simpler, better compiler-optimized, and thread-safe.

**Comment by feldergast** [comment]
I understand how Meyers singletons work and don't doubt they are better than what was there.  Those pieces of code are not what I was actually thinking about.  Looking back through the PR, it looks like you changed the trigger mechanism, but not with a Meyers singleton. This is the line I was thinking about in the new code:

static inline const bool loaded = Base::Ctor::template add<T>();

which replaced this code defined outside of the class (static bool loaded was declared in the class):

template <class Base, class T>
const bool InstantiateBuilder<Base, T>::loaded = Base::Ctor::template add<T>();

So not a Meyers singleton, just a newer/cleaner way to force the variable to initialize on library load.  I don't think this was allowed in C++11, which is what the original code was written in. I know the old standard did not have a firm definition of how this type of initialization should function when loading multiple libraries with the same template expansion.  If I remember correctly, the Intel compiler treated this differently than other compilers.  Is it better defined in the standard how this type of statement should interact across libraries? I think the Intel compiler overwrote the ELI database that was statically initialized by previously opened libraries instead of using what was already there.  There is a whole layer to templating that was put in to fix this issue up by triggering a reload of everything every time a new library loaded (at least when using the Intel compiler).

**Comment by leekillough** [comment]
I think we are talking about different things. I am not talking about that `loaded` variable. That change made it `inline` so that it could be defined inside the class and multiple definitions could occur in header files without violating the One Definition Rule (ODR), but moving it into an `inline` variable defined inside the class does not change its initialization order w.r.t. other global variables.

I am talking about changes like this (this is only one, but I think there were more):

```Diff
-    static void add(const std::string& elemlib, const std::string& elem, T* info)
-    {
-        if ( !infos_ ) {
-            infos_ = std::unique_ptr<std::map<std::string, std::map<std::string, T*>>>(
-                new std::map<std::string, std::map<std::string, T*>>);
-        }
-
-        (*infos_)[elemlib][elem] = info;
-    }
+    static void add(const std::string& elemlib, const std::string& elem, T* info) { infos()[elemlib][elem] = info; }

 private:
-    static std::unique_ptr<std::map<std::string, std::map<std::string, T*>>> infos_;
+    static std::map<std::string, std::map<std::string, T*>>& infos()
+    {
+        static std::map<std::string, std::map<std::string, T*>> infos_;
+        return infos_;
+    }
 };

```

Here, `infos_` was a `std::unique_ptr` which was allocated the first time `add()` was called. I changed it so that `infos()` function returns a Meyers singleton which is then used in the `add()` function. There are other examples of changes of mine like this in 86ef30fb29 and other commits, like this:
```Diff
     static Library* getLibrary(const std::string& name)
     {
-        if ( !libraries ) { libraries = new std::map<std::string, Library*>; }
-        auto iter = libraries->find(name);
-        if ( iter == libraries->end() ) {
-            auto* info         = new Library(name);
-            (*libraries)[name] = info;
-            return info;
-        }
-        else {
-            return iter->second;
-        }
+        static Map libraries; // Database
+        auto&      lib = libraries[name];
+        if ( !lib ) lib = new Library(name);
+        return lib;
     }

     template <class NewBase>
     using ChangeBase = BuilderLibraryDatabase<NewBase, CtorArgs...>;
-
-private:
-    // Database - needs to be a pointer for static init order
-    static Map* libraries;

```
Here, contrary to the comment which says `needs to be a pointer for static init order`, the `libraries` variable can be a `static` variable _inside_ the `getLibrary()` function which gets initialized on the first call, without needing to call `new`. 
```Diff
@@ -168,14 +145,10 @@ struct CachedAllocator
     template <class... Args>
     Base* operator()(Args&&... ctorArgs)
     {
-        if ( !cached_ ) { cached_ = new T(std::forward<Args>(ctorArgs)...); }
-        return cached_;
+        static T* cached = new T(std::forward<Args>(ctorArgs)...);
+        return cached;
     }
-
-    static Base* cached_;
 };
-template <class Base, class T>
-Base* CachedAllocator<Base, T>::cached_ = nullptr;
```
Another one where `cached_` used to be defined in the class and initialized to `nullptr`, and was checked for `nullptr` on every call to `operator()` but didn't need that -- simply making it a `static` variable _inside of the function_ makes it get initialized on the first call. Here it seems to need to be a pointer although it's possible it could just be a static variable instead of a pointer.

Previously, they had pointers and/or `std:unique_ptr` which were allocated on the first call if they were `nullptr`, which I thought the submitter of this issue was talking about -- memory leaks of ELI pointers allocated for static variables. By moving them to become static variables **inside** functions which get initialized on the first call, **it makes the pointer leak issue moot**.

**Comment by feldergast** [comment]
Interesting discussion.  I agree that the spirit of this issue was meet with #1222 and we should close this issue.  I would like to summarize what I think I learned from this conversation and from looking up some of the concepts described here since they are key concepts to how the ELI code works:

1 - Initialization of static variables defined within a function happens on first invocation of that function.  This is thread-safe and is known as the Meyers Singleton. This works as long as all access to the variable goes through the function.

2 - Initialization of static variables at class or global scope happens either at program startup, or when the dynamic library is loaded.  These are subject to the [static initialization order fiasco](https://isocpp.org/wiki/faq/ctors#static-init-order).  I'm not sure this applies to the ELI as none of the ELI data structures call into other structures of different types, so initialization order isn't important.  There are some shared resources, but since initialization is serial (library loading is mutexed), it doesn't matter which data structure first initializes those. Many of these shared allocations were switched to use Meyers Singletons for added safety and to better manage leaking of pointers.

3 - The inline keyword can now be used with static variables initialized in class to guarantee that the [One Definition Rule](https://en.cppreference.com/w/cpp/language/definition) is followed.  The linked page seems to indicate that you are actually allowed to have more than one definition of templated variables (with several caveats, that I think are met), so I'm not sure the in-class inlined static initialization of the loaded variable is strictly required, but it is definitely a cleaner way to do things with better guarantees.

**Comment by leekillough** [comment]
Your summary conclusions are correct, and well-summarize what I was trying to say.

As for #3, yes, `inline` functions and C++17 `inline` variables (including `static inline` and `static constexpr` class member variables), and `template` functions and C++14 `template` variables, are allowed to be defined in multiple compilation units as long as the definitions are the same "sequence of tokens" or words to that effect. However, for these purposes, a [full specialization](https://en.cppreference.com/w/cpp/language/template_specialization) of a function or variable template (syntax `template<>` ...)  is not considered "templated" anymore, so it can cause multiple definition errors if it's not explicitly marked `inline`, and the `inline` specifier on the primary template definition does not imply `inline` on specializations. But templated overloads of a function, or the primary variable template and any _partial_ specializations thereof, do not need to be explicitly marked `inline` to avoid ODR violations. Function templates cannot have partial specializations, only full specializations and multiple templated overloads.
